wdl-cli/src/inputs.rs

```diff
@@ -0,0 +1,507 @@
+                self.extend(inputs.into_inner());
+            }
+            Input::Pair { key, value } => {
+                // SAFETY: we expect that the current working directory is
+                // always available for the platforms that `wdl` will run
+                // within.
+                let cwd = std::env::current_dir().unwrap();
+                self.insert(key, (cwd, value));
+            }
+        };
+
+        Ok(())
+    }
+
+    /// Attempts to coalesce a set of inputs into an [`Inputs`].
+    pub fn coalesce<T, V>(iter: T) -> Result<Self>
+    where
+        T: IntoIterator<Item = V>,
+        V: AsRef<str>,
+    {
+        let mut inputs = Inputs::default();
+
+        for input in iter {
+            inputs.add_input(input.as_ref())?;
+        }
+
+        Ok(inputs)
+    }
+
+    /// Consumes `self` and returns the inner index map.
+    pub fn into_inner(self) -> InputsInner {
+        self.0
+    }
+
+    /// Converts a set of inputs to a set of engine inputs.
+    ///
+    /// Returns `Ok(Some(_))` if the inputs are not empty.
+    ///
+    /// Returns `Ok(None)` if the inputs are empty.
+    ///
+    /// When the inputs are not empty, the return type contained in `Some(_)` is
+    /// a tuple of,
+    ///
+    /// - the name of the callee (the name of the task or workflow being run),
+    /// - the transformed engine inputs, and
+    /// - a map containing the origin path for each provided input key.
+    pub fn into_engine_inputs(
+        self,
+        document: &Document,
+    ) -> anyhow::Result<Option<(String, EngineInputs, OriginPaths)>> {
+        let (origins, values): (IndexMap<_, _>, IndexMap<_, _>) = self
+            .0
+            .into_iter()
+            .map(|(key, (origin, value))| ((key.clone(), origin), (key, value)))
+            .unzip();
+
+        let object = Object::from(values);
+        let result = EngineInputs::parse_object(document, object)?;
+
+        Ok(result.map(|(callee_name, inputs)| {
+            let callee_prefix = format!("{}.", callee_name);
+
+            let origins = origins
+                .into_iter()
+                .map(|(key, path)| {
+                    if let Some(key) = key.strip_prefix(&callee_prefix) {
+                        (key.to_owned(), path)
+                    } else {
+                        (key, path)
+                    }
+                })
+                .collect::<IndexMap<String, PathBuf>>();
+
+            (callee_name, inputs, OriginPaths::from(origins))
+        }))
+    }
+}
+
+impl Deref for Inputs {
+    type Target = InputsInner;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for Inputs {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn identifier_regex() {
+        assert!(IDENTIFIER_REGEX.is_match("here_is-an.identifier"));
```

Comment:
Are dashes allowed in WDL identifiers? I don't think they are, but I could be wrong.